packages/grafana-ui/src/components/DateTimePickers/TimeRangeContext.tsx

```diff
@@ -0,0 +1,104 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import { TimeRange } from '@grafana/data';
+
+type TimeRangeContextValue = TimeRangeContextHookValue & {
+  // These are to be used internally and aren't passed to the users of the hook.
+
+  // Called when picker is mounted to update the picker count.
+  addPicker(): void;
+
+  // Called when picker is unmounted to update the picker count.
+  removePicker(): void;
+};
+
+export type TimeRangeContextHookValue = {
+  // If the time range is synced, this is the value that all pickers should show.
+  syncedValue?: TimeRange;
+
+  // If the time range is synced across all visible pickers.
+  synced: boolean;
+
+  // If it is possible to sync the time range. This is based just on the number of pickers that are visible so if
+  // there is only one, the sync button can be hidden.
+  syncPossible: boolean;
+
+  // Action passed to the picker to interact with the sync state.
+  // Sync the time range across all pickers with the provided value. Can be also used just to update a value when
+  // already synced.
+  sync(value: TimeRange): void;
+  unSync(): void;
+};
+
+const TimeRangeContext = React.createContext<TimeRangeContextValue | undefined>(undefined);
+
+export function TimeRangeProvider({ children }: { children: React.ReactNode }) {
+  // We simply keep the count of the pickers visible by letting them call the addPicker and removePicker functions.
+  const [pickersCount, setPickersCount] = useState(0);
+  const [synced, setSynced] = useState(false);
+  const [syncedValue, setSyncedValue] = useState<TimeRange>();
```

Comment:
Can we infer the synced state from syncedValue? When it's undefined we can consider the pickers unsynced, opposite otherwise. So far it seems redundant to have both of them.

Reply:
You are right. I removed the `synced` state variable, but I kept the `synced` context prop as `Boolean(syncedValue)`. To me it seemed like more understandable, like it's a clear name of state and you don't have to sort of transform `Boolean(syncedValue)` yourself but let me know if you think that makes sense, don't have a super strong feeling towards it, compared to just having `syncedValue` in the context object and let users deal with it.

Reply:
Yeah, I think exposing `synced` as a boolean makes sense 👍